File: admin-dashboard/src/api/chats.ts
/**
 * Chats API client and response schemas.
 */

import { z } from "zod";

import { request, type ApiResult } from "./client";

export const ChatConversationItemSchema = z.object({
  telegramUserId: z.number(),
  username: z.string().nullable(),
  firstName: z.string().nullable(),
  lastName: z.string().nullable(),
  phoneE164: z.string().nullable(),
  hasAvatar: z.boolean(),
  lastMessageText: z.string().nullable(),
  lastMessageDirection: z.enum(["inbound", "outbound"]).nullable(),
  lastMessageKind: z.string().nullable(),
  lastMessageAt: z.string(),
  messageCount: z.number(),
});

export type ChatConversationItem = z.infer<typeof ChatConversationItemSchema>;

export const ChatConversationsResponseSchema = z.object({
  items: z.array(ChatConversationItemSchema),
  total: z.number(),
});

export type ChatConversationsResponse = z.infer<typeof ChatConversationsResponseSchema>;

export const ChatMessageViewSchema = z.object({
  id: z.string(),
  telegramUserId: z.number(),
  chatId: z.number(),
  direction: z.enum(["inbound", "outbound"]),
  kind: z.enum(["text", "callback", "screen", "audio", "voice", "toast", "action"]),
  body: z.string().nullable(),
  wizardStep: z.string().nullable(),
  telegramMessageId: z.number().nullable(),
  callbackData: z.string().nullable(),
  isTruncated: z.boolean(),
  createdAt: z.string(),
});

export type ChatMessageView = z.infer<typeof ChatMessageViewSchema>;

export const ChatTranscriptResponseSchema = z.object({
  telegramUserId: z.number(),
  messages: z.array(ChatMessageViewSchema),
  total: z.number(),
});

export type ChatTranscriptResponse = z.infer<typeof ChatTranscriptResponseSchema>;

export interface FetchConversationsParams {
  readonly q?: string;
  readonly limit?: number;
  readonly offset?: number;
  readonly signal?: AbortSignal;
}

export async function fetchConversations(
  params: FetchConversationsParams = {},
): Promise<ApiResult<ChatConversationsResponse>> {
  const query = new URLSearchParams();
  if (params.q?.trim()) query.set("q", params.q.trim());
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));

  const qs = query.toString();
  const path = qs ? `/api/chats?${qs}` : "/api/chats";

  return request({
    endpoint: "GET /api/chats",
    path,
    schema: ChatConversationsResponseSchema,
    ...(params.signal ? { signal: params.signal } : {}),
  });
}

export interface FetchTranscriptParams {
  readonly limit?: number;
  readonly beforeId?: string;
  readonly signal?: AbortSignal;
}

export async function fetchChatTranscript(
  telegramUserId: number,
  params: FetchTranscriptParams = {},
): Promise<ApiResult<ChatTranscriptResponse>> {
  const query = new URLSearchParams();
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.beforeId) query.set("before_id", params.beforeId);

  const qs = query.toString();
  const path = qs
    ? `/api/chats/${encodeURIComponent(String(telegramUserId))}/messages?${qs}`
    : `/api/chats/${encodeURIComponent(String(telegramUserId))}/messages`;

  return request({
    endpoint: "GET /api/chats/{telegram_user_id}/messages",
    path,
    schema: ChatTranscriptResponseSchema,
    ...(params.signal ? { signal: params.signal } : {}),
  });
}

File: src/hbd/admin/schemas/chats.py
"""Wire schemas for the Chats console.

ADMIN_PANEL_PLAN §5.7, §6.7.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from hbd.admin.schemas.common import ApiModel
from hbd.db.enums import ChatDirection, ChatMessageKind

__all__ = [
    "ChatConversationItem",
    "ChatConversationsResponse",
    "ChatMessageView",
    "ChatTranscriptResponse",
]


class ChatConversationItem(ApiModel):
    """Summary of one customer conversation thread."""

    telegram_user_id: int
    chat_id: int
    user_id: UUID | None
    username: str | None
    first_name: str | None
    last_name: str | None
    phone_e164: str | None
    has_avatar: bool
    is_blocked: bool
    last_message_at: datetime
    last_message_text: str | None
    last_message_direction: ChatDirection
    last_message_kind: ChatMessageKind
    message_count: int
    wizard_step: str | None


class ChatConversationsResponse(ApiModel):
    """GET /api/chats response envelope."""

    items: list[ChatConversationItem]
    total: int


class ChatMessageView(ApiModel):
    """One message within a customer chat transcript."""

    id: UUID
    telegram_user_id: int
    chat_id: int
    order_id: UUID | None
    session_id: str | None
    direction: ChatDirection
    kind: ChatMessageKind
    wizard_step: str | None
    telegram_message_id: int | None
    callback_data: str | None
    parse_mode: str | None
    body: str | None
    is_truncated: bool
    is_from_worker: bool
    media_group_id: str | None
    error_code: str | None
    correlation_id: str | None
    created_at: datetime


class ChatTranscriptResponse(ApiModel):
    """GET /api/chats/{telegram_user_id}/messages response envelope."""

    telegram_user_id: int
    messages: list[ChatMessageView]
    total: int
